tools/asset-sync/assets.ps1

```diff
@@ -0,0 +1,521 @@
+
+    # this is where we will need to handle the multi-copying of the repository.
+    # to begin with, we will use the relative path of the assets json in combination with the
+    # Repo/RepoId to create a unique hash. In the future, we will need to take the targeted commit into account,
+    # and resolve conflicts
+    if ($Config.AssetsRepoId) {
+        $repoName = $Config.AssetsRepoId
+    }
+
+    $repoNameHashed = Get-MD5-Hash -Input ((Join-Path $repoName $Config.AssetsJsonRelativeLocation).ToString())
+
+    $repoPath = (Join-Path $assetsLocation $repoNameHashed.Hash)
+    
+    if (-not (Test-Path $repoPath)){
+        New-Item -Type Directory -Force -Path $repoPath | Out-Null
+    }
+    
+    return $repoPath
+}
+
+Function Get-Default-Branch {
+    param(
+        $Config
+    )
+    $repoJsonResult = curl "https://api.github.com/repos/$($Config.AssetsRepo)"
+    return ($repoJsonResult | ConvertFrom-Json).default_branch
+}
+
+<#
+.SYNOPSIS
+This function returns a boolean that indicates whether or not the assets repo has been initialized.
+#>
+Function Is-AssetsRepo-Initialized {
+    param(
+        [PSCustomObject] $Config
+    )
+
+    $result = $false
+    $assetRepoLocation = Resolve-AssetRepo-Location -Config $Config
+
+    try {
+        Push-Location $assetRepoLocation
+        $gitLocation = Join-Path $assetRepoLocation ".git"
+
+        if (Test-Path $gitLocation){
+            $result = $true
+        }
+        else{
+            $result = $false
+        }
+    }
+    catch {
+        Write-Error $_
+        $result = $false
+    }
+    finally {
+        Pop-Location
+    }
+
+    return $result
+}
+
+
+<#
+.SYNOPSIS
+Given a configuration, determine which paths must be added to the sparse checkout of the assets repo.
+
+.PARAMETER Config
+A PSCustomObject that contains an auto-parsed assets.json content.
+
+#>
+Function Resolve-CheckoutPaths {
+    param(
+        [Parameter(Mandatory=$true)]
+        [PSCustomObject] $Config
+    )
+    $assetsJsonPath = $Config.AssetsJsonRelativeLocation
+
+    $assetsJsonFolder = Split-Path $Config.AssetsJsonRelativeLocation
+
+    if(!$assetsJsonFolder -or $assetsJsonFolder -in @(".", "./", ".\"))
+    {
+        return $Config.AssetsRepoPrefixPath.Replace("`\", "/")
+    }
+    else {
+        $result = (Join-Path $Config.AssetsRepoPrefixPath $assetsJsonFolder).Replace("`\", "/")
+        return $result
+    }
+}
+
+<#
+.SYNOPSIS
+Given a configuration, determine the _current_ target path.
+
+.DESCRIPTION 
+Determines the presence of a branch on the git repo. If the relevant auto/<service> branch does not exist, we should use main.
+#>
+Function Resolve-TargetBranch {
+    param(
+        [Parameter(Mandatory=$true)]
+        [PSCustomObject] $Config
+    )
+    $assetRepo = Resolve-AssetRepo-Location -Config $Config
+    $branch = "main"
+    try {
+        Push-Location $assetRepo
+        
+        $latestCommit = git rev-parse "origin/$($Config.AssetsRepoBranch)"
```

Comment:
If the branch doesn't exist, you get exception text on stdErr:

```
fatal: ambiguous argument 'origin/auto/test': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
origin/auto/test
```

To disambiguate and silence the error text:
```
$latestCommit = git rev-parse "origin/$($Config.AssetsRepoBranch)" -- 2>$null
```

Reply:
It's necessary to disambiguate because someone could have the local relative path `origin/main`. If they do, `git rev-parse origin/main` returns `origin/main`

Reply:
It's necessary to disambiguate. If the user has the local relative path `origin/auto/something` and not the ref, `git rev-parse origin/auto/something` will return the local relative path and not a commit id or 128 exit code.

Reply:
Done. Thanks patrick!